hw/misc/pcie-testdev.c

```diff
@@ -0,0 +1,202 @@
+    s->regs.status = 0x01;
+    printf("DMA Transfer from source addr %p to dest addr %p, length %lu\n", (void *)s->regs.srcaddr, (void *)s->regs.dstaddr, s->regs.size);
+    uint16_t *buf = g_malloc(s->regs.size);
+    printf("Read from source, STATUS: %d\n", pci_dma_read(&s->parent_obj, s->regs.srcaddr, buf, s->regs.size));
+    printf("Write to dest, STATUS: %d\n", pci_dma_write(&s->parent_obj, s->regs.dstaddr, buf, s->regs.size));
+    g_free(buf);
+    s->regs.status = 0x00;
+
+    if (s->aggressive) {
+        QEMUTimer *dma_timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) dma_transfer, s);
+        timer_mod(dma_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + s->aggression_delay);
+    }
+}
+
+static void
+pcie_testdev_write(void *opaque, hwaddr addr, uint64_t val,
+                   unsigned size)
+{
+    PCIeTestDevState *s = opaque;
+
+    switch(addr) {
+    case SRCADDR:
+        s->regs.srcaddr = val;
+        break;
+    case DSTADDR:
+        s->regs.dstaddr = val;
+        break;
+    case SIZE:
+        s->regs.size = val;
+        break;
+    case DOORBELL:
+        s->regs.doorbell = val;
+        if (val & 0x01 && s->regs.size) {
+            s->regs.status = 0x01;
+
+            /* Use a timer to simulate delay in device making a DMA_transfer */
+            QEMUTimer *timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) dma_transfer, s);
+            timer_mod(timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + s->regs.busy_delay);
+        }
+        break;
+    case BUSY_DELAY:
+        s->regs.busy_delay = val;
+        break;
+    default:
+        break;
+    }
+}
+
+static const MemoryRegionOps pcie_testdev_io_ops = {
+    .read = pcie_testdev_read,
+    .write = pcie_testdev_write,
+    .endianness = DEVICE_LITTLE_ENDIAN,
+};
+
+static void pcie_testdev_realize(PCIDevice *pcie_dev, Error **errp)
+{
+    PCIeTestDevState *s = PCI_EXPRESS_TEST_DEV(pcie_dev);
+
+    memory_region_init_io(&s->membar, OBJECT(s), &pcie_testdev_io_ops, s,"pci-testdev-membar", 65536);
+    pci_register_bar(pcie_dev, 0, PCI_BASE_ADDRESS_SPACE_MEMORY|
+                     PCI_BASE_ADDRESS_MEM_PREFETCH |
+                     PCI_BASE_ADDRESS_MEM_TYPE_64, &s->membar);
+
+    pcie_endpoint_cap_init(pcie_dev, 0);
+
+    /* Aggressor mode: Self-enable bus mastering and start DMA immediately */
+    if (s->aggressive) {
+        QEMUTimer *bus_master_timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) enable_bus_master, s);
+        timer_mod(bus_master_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL));
+        QEMUTimer *dma_timer = timer_new_ms(QEMU_CLOCK_VIRTUAL, (void *) dma_transfer, s);
+        timer_mod(dma_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL));
+    }
+}
+
+static void pcie_testdev_class_init(ObjectClass *klass, void *data)
+{
+    DeviceClass *dc = DEVICE_CLASS(klass);
+    PCIDeviceClass *k = PCI_DEVICE_CLASS(klass);
+
+    k->device_id = PCI_EXPRESS_DEVICE_ID_TEST;
+    k->vendor_id = PCI_VENDOR_ID_REDHAT;
```

Comment:
Perhaps we should use Google (0x1ae0) here?

I can reserve a device ID.

Reply:
Switched to Google vendor ID and allocated device and subsystem IDs